R/ReportDocument.R

```diff
@@ -0,0 +1,200 @@
+  # Regular c() drops classes and attributes, so we either overwrite the method
+  # or we do not use ReportDocument class, but list class.
+
+  # Does not work, if ReportDocument is the second element, and not the first.
+  # teal.reporter::report_document() -> x
+  # class(c(list(), x)) # list
+  # class(c(x, list())) # ReportDocument
+  # append(x, list(), after = 1) # ReportDocument
+  # append(x, list(), after = 0) # list()
+
+  input_objects <- list(...)
+  attrs <- attributes(input_objects[[1]])
+  objects <- do.call(c, lapply(input_objects, unclass))
+  attributes(objects) <- attrs
+  objects
+}
+
+#' @rdname report_document
+#' @export
+`[.ReportDocument` <- function(x, i) {
+  # Regular [] drops classes, so we either overwrite the method
+  # or we do not use ReportDocument class, but list class.
+  attrs <- attributes(x)
+  xi <- unclass(x)[i]
+  attributes(xi) <- attrs
+  xi
+}
+
+#' @rdname report_document
+#' @param x `ReportDocument`
+#' @param modify `integer(n)` if present, uses `[.` syntax to extract elements.
+#' Can be used to reorder or substract the object
+#' @param append object to be appended to `ReportDocument` with `append` syntax.
+#' Use `after` to specify the position where the object should be added.
+#'
+#' @examples
+#' report <- report_document(1, 2, 'c')
+#'
+#' # Modify and append to the report
+#' new_report <- edit_document_content(report, modify = c(3, 1), append = 'd')
+#' new_report
+#' class(new_report)
+#'
+#' @export
+edit_document_content <- function(x, modify = NULL, append = NULL, after = length(x)) {
+  checkmate::assert_class(x, "ReportDocument")
+  checkmate::assert_class(modify, "numeric", null.ok = TRUE)
+
+  attrs <- attributes(x)
+
+  if (!is.null(modify)) {
+    x <- x[modify]
+  }
+
+  if (!is.null(append)) {
+    x <- append(x, append, after)
+  }
+
+  attributes(x) <- attrs
+  x
+
+}
+
+#' Generate an R Markdown code chunk
+#'
+#' This function takes a character string as input and formats it as an R Markdown code chunk.
+#' Additional named parameters passed via `...` will be included inside `{r}`.
+#'
+#' @param code A character string containing the R code to be wrapped in the chunk.
+#' @param ... Additional named parameters to be included inside `{r}`.
+#'
+#' @return A formatted character string representing an R Markdown code chunk.
+#' @examples
+#' code_chunk("x <- 1:10", echo = TRUE, message = FALSE)
+#' @export
+#' @rdname code_output
+code_chunk <- function(code, ...) {
+  params <- list(...)
+  params_str <- if (length(params) > 0) {
+    paste(names(params), params, sep = "=", collapse = ", ")
+  } else {
+    ""
+  }
+
+  if(!grepl("eval=", params_str, fixed = TRUE)) {
+    if (params_str == "") {
+      params_str <- "eval=FALSE"
+    } else {
+      params_str <- paste0(params_str, ", eval=FALSE")
+    }
+  }
+  code_chunk_id <- paste0('code_chunk_', rlang::hash(code))
+  sprintf("```{r %s, %s}\n%s\n```", code_chunk_id, params_str, code)
```

Comment:
Wrapping in ticks here causes problems later. Class should be rendered to desired output without need to parse-replace text.

```suggestion
  params <- list(...)
  structure(
    code,
    params = params,
    class = "code_chunk"
  )
```

Reply:
This issue is likely related to `clean_chunks` as well. https://github.com/insightsengineering/teal.reporter/pull/307#discussion_r2011678924

Currently, the process of generating a report output is a bit messy and works as follows:  

1. Retrieve objects from the reporter:  
   - If the object is plain text, it's directly inserted into the `.Rmd` file.  
   - If the object is an R object, it is saved as an `.Rds` file, and a corresponding chunk is added to load it.  
   - No code chunks are executed, except for those responsible for loading `.Rds` files.  

2. Render the `.Rmd` file:  
   - During rendering, the chunks that load `.Rds` files are evaluated, but afterward, they are removed from the final `.Rmd` file.  

This approach ensures that the final `.Rmd` file doesn't contain the code that loads `.Rds` objects while still producing a report output with the expected results—without needing to rerun the code.  

Since the code might not always be reproducible, I ensure that all chunks remain unevaluated during rendering (except for library imports, since those might be necessary for `print.object_class` functions). However, chunks that load saved `.Rds` files are evaluated. Later, I clean up the file accordingly.  

I believe we can delay the process of adding backticks (`` ` ``) until a later stage, like during the `block_to_markdown` phase. Then, in `block_to_html`, we can apply formatting based on a specific class if we create a structured class for this purpose.  

After thinking about it for a while, I’m convinced this approach makes a lot of sense.

Reply:
Alrighty, now using a simple `code_chunk()` function that hold the code and parameters. It is then handled by `block_to_html` (wrapped in `tags$pre` tag) and `block_to_md` (adding ticks, and if output is powerpoint doing something more).